lib/src/arithmetic/divider.dart

```diff
@@ -0,0 +1,362 @@
+
+  /// Request for a new division operation to be performed.
+  Logic get validIn => port('validIn');
+
+  /// Quotient (result) for the division operation.
+  Logic get quotient => port('quotient');
+
+  /// A Division by zero occurred.
+  Logic get divZero => port('divZero');
+
+  /// The result of the current division operation is ready.
+  Logic get validOut => port('validOut');
+
+  /// The module is ready to accept new inputs.
+  Logic get readyIn => port('readyIn');
+
+  /// The width of the data operands and result.
+  final int dataWidth;
+
+  /// A constructor for the divider interface.
+  MultiCycleDividerInterface({this.dataWidth = 32})
+      : super(portsFromProvider: [
+          Port('clk'),
+          Port('reset'),
+          Port('dividend', dataWidth),
+          Port('divisor', dataWidth),
+          Port('validIn'),
+          Port('readyOut'),
+        ], portsFromConsumer: [
+          Port('quotient', dataWidth),
+          Port('divZero'),
+          Port('validOut'),
+          Port('readyIn'),
+        ]);
+
+  /// A match constructor for the divider interface.
+  MultiCycleDividerInterface.match(MultiCycleDividerInterface other)
+      : this(dataWidth: other.dataWidth);
+}
+
+/// The Divider module definition
+class MultiCycleDivider extends Module {
+  /// The Divider's interface declaration.
+  late final MultiCycleDividerInterface intf;
+
+  /// Get interface's validOut signal value.
+  Logic get validOut => intf.validOut;
+
+  /// Get interface's quotient signal value.
+  Logic get quotient => intf.quotient;
+
+  /// Get interface's divZero signal value.
+  Logic get divZero => intf.divZero;
+
+  /// Get interface's readyIn signal value.
+  Logic get readyIn => intf.readyIn;
+
+  /// The width of the data operands and result.
+  late final int dataWidth;
+
+  /// The log of the data width representing
+  /// the number of bits required to store that number.
+  late final int logDataWidth;
+
+  /// The Divider module's constructor
+  MultiCycleDivider(MultiCycleDividerInterface interface)
+      : dataWidth = interface.dataWidth,
+        logDataWidth = log2Ceil(interface.dataWidth),
+        super(name: 'divider') {
+    intf = MultiCycleDividerInterface.match(interface)
+      ..pairConnectIO(
+        this,
+        interface,
+        PairRole.consumer,
+        uniquify: (original) => '${super.name}_$original',
+      );
+
+    _build();
+  }
+
+  /// Named constructor that explicitly takes Logics instead of an Interface.
+  MultiCycleDivider.ofLogics({
+    required Logic clk,
+    required Logic reset,
+    required Logic validIn,
+    required Logic dividend,
+    required Logic divisor,
+    required Logic readyOut,
+  })  : assert(dividend.width == divisor.width,
+            'Widths of all data signals do not match!'),
+        dataWidth = dividend.width,
+        logDataWidth = log2Ceil(dividend.width),
+        super(name: 'divider') {
+    intf = MultiCycleDividerInterface(dataWidth: dataWidth);
```

Comment:
I would change this constructor to a `factory` constructor that then passes this interface into a call to the default constructor.  This interface is not actually creating any ports on the module

It would be a good idea to add a test that this constructor at least builds the module correctly (you could check that ports are created for example).  No need to retest functionality with this constructor.

Reply:
Is there special syntax for factory constructors in Dart? Or is it fair to just export a factory function outside of the class entirely?

Reply:
Check out the factory constructors in here for reference:
https://github.com/mkorbel1/rohd-hcl/blob/counter/lib/src/summation/counter.dart